src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1577,7 +1577,7 @@ ICollection IDictionary.Values
                 if (key == null) ThrowKeyNullException();
 
                 if (!(key is TKey)) throw new ArgumentException(SR.ConcurrentDictionary_TypeOfKeyIncorrect);
-                if (!(value is TValue)) throw new ArgumentException(SR.ConcurrentDictionary_TypeOfValueIncorrect);
+                if (!(value is TValue) && value != null) throw new ArgumentException(SR.ConcurrentDictionary_TypeOfValueIncorrect);
```

Comment:
Should this instead do the same validation logic as in Add?  I'm wondering why they're different.
https://github.com/dotnet/corefx/pull/26448/files#diff-f1285172a2a161542ccf657778f0852cL1438

This is also a different check than Dictionary does:
https://github.com/dotnet/coreclr/blob/d44f3e883a803b2d1fe4fb2c07f92d78bff56eed/src/mscorlib/src/System/ThrowHelper.cs#L355

Reply:
I think the logic should be the same in both routines, in the add method and in the indexer's setter. There is one more difference in the ConcurrentDictionary class. In the example ConcurrentDictionary.Add throws a NullReference exception (is it a desired behavior or another bug?) and ConcurrentDictionary.Indexer.Setter throws an Argument exception.  

And in the Dictionary class both routines act the same.

```c#
// ConcurrentDictionary using IDictionary
{
    IDictionary d = new ConcurrentDictionary<string, int>();
    
    // NullReferenceException - Object reference not set to an instance of an object.
    d.Add("key", null); 
}

{
    IDictionary d = new ConcurrentDictionary<string, int>();
    
    // ArgumentException - The value was of an incorrect type for this dictionary.
    d["key"] = null; 
}

// Dictionary using IDictionary
{
    IDictionary d = new Dictionary<string, int>();

    // ArgumentNullException - Value cannot be null. Parameter name: value
    d.Add("key", null); 
}

{
    IDictionary d = new Dictionary<string, int>();

    // ArgumentNullException - Value cannot be null. Parameter name: value
    d["key"] = null; 
}
```

Reply:
We're fine with the fix and it is definitely a bug, but we should use either of both validations that @stephentoub pointed out to, since those consider ValueType objects which are not nullable.